hypergan/graph/graph.py

```diff
@@ -0,0 +1,335 @@
+        def split(layer):
+            start = 0
+            ret = []
+            for category in categories:
+                count = int(category.get_shape()[1])
+                ret.append(tf.slice(layer, [0, start], [batch_size, count]))
+                start += count
+            return ret
+
+        for category,layer_s in zip(categories, split(layer)):
+            size = int(category.get_shape()[1])
+            category_prior = tf.ones([batch_size, size])*np.float32(1./size)
+            logli_prior = tf.reduce_sum(tf.log(category_prior + TINY) * category, reduction_indices=1)
+            layer_softmax = tf.nn.softmax(layer_s)
+            logli = tf.reduce_sum(tf.log(layer_softmax+TINY)*category, reduction_indices=1)
+            disc_ent = tf.reduce_mean(-logli_prior)
+            disc_cross_ent =  tf.reduce_mean(-logli)
+
+            loss += disc_ent - disc_cross_ent
+        return loss
+
+    def random_category(batch_size, size, dtype):
+        prior = tf.ones([batch_size, size])*1./size
+        dist = tf.log(prior + TINY)
+        with tf.device('/cpu:0'):
+            sample=tf.multinomial(dist, num_samples=1)[:, 0]
+            return tf.one_hot(sample, size, dtype=dtype)
+
+
+    # Used for building the tensorflow graph with only G
+    def create_generator(self, x,y,f):
+        config = self.config
+        set_ops_globals(config['dtype'], config['batch_size'])
+        z_dim = int(config['generator.z'])
+        z, encoded_z, z_mu, z_sigma = config['encoder'](config, x, y)
+        categories = [random_category(config['batch_size'], size, config['dtype']) for size in config['categories']]
+        if(len(categories) > 0):
+            categories_t = [tf.concat(1, categories)]
+        else:
+            categories_t = []
+
+
+        args = [y, z]+categories_t
+        g = self.generator(args)
+        set_tensor("g", g)
+        set_tensor("y", y)
+        set_tensor('categories', categories_t)
+
+        return g
+
+    def create(self, x,y,f):
+        config = self.config
+        # This is a hack to set dtype across ops.py, since each tensorflow instruction needs a dtype argument
+        # TODO refactor
+        set_ops_globals(config['dtype'], config['batch_size'])
+
+        batch_size = config["batch_size"]
+        z_dim = int(config['generator.z'])
+        batch_norm = config['generator.regularizers.layer']
+
+        g_losses = []
+        extra_g_loss = []
+        d_losses = []
+
+        #initialize with random categories
+        categories = [random_category(config['batch_size'], size, config['dtype']) for size in config['categories']]
+        if(len(categories) > 0):
+            categories_t = [tf.concat(1, categories)]
+        else:
+            categories_t = []
+
+        z, encoded_z, z_mu, z_sigma = config['encoder'](config, x, y)
+
+        # create generator
+        g = self.generator([y, z]+categories_t)
+
+        #encoded = generator([y, encoded_z]+categories_t, reuse=True)
```

Comment:
do we want to support autoencoding in this version or just cut it out of the base?  Would it make sense as an api example?